akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -491,24 +490,41 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
         None // don't change state
       }
     }
+    def onTrailingHeaders(headers: ParsedHeadersFrame): Unit = {
+      trailingHeaders = Some(headers)
+      if (headers.endStream) {
+        onDataFrame(DataFrame(headers.streamId, endStream = true, ByteString.empty)) // simulate end stream by empty dataframe
+      } else pushGOAWAY(Http2Protocol.ErrorCode.PROTOCOL_ERROR, "Got unexpected mid-stream HEADERS frame")
+    }
     def onRstStreamFrame(rst: RstStreamFrame): Unit = {
       outlet.fail(new PeerClosedStreamException(rst.streamId, rst.errorCode))
       buffer = ByteString.empty
+      trailingHeaders = None
       wasClosed = true
     }
 
     private def dispatchNextChunk(): Unit = {
       if (buffer.nonEmpty && outlet.isAvailable) {
         val dataSize = buffer.size min settings.requestEntityChunkSize
-        outlet.push(buffer.take(dataSize))
+        outlet.push(wrapData(buffer.take(dataSize)))
         buffer = buffer.drop(dataSize)
 
         totalBufferedData -= dataSize
 
         debug(s"Dispatched chunk of $dataSize for stream [$streamId], remaining window space now $outstandingStreamWindow, buffered: ${buffer.size}")
         updateWindows()
       }
-      if (buffer.isEmpty && wasClosed) outlet.complete()
+      if (buffer.isEmpty && wasClosed) {
```

Comment:
Shouldn't this also have a `outlet.isAvailable` condition like the block above? The block above mutates the buffer so if it contained one last chunk it could try to do a second push with the trailing headers here (which would throw).

What triggers `outlet.complete()` if there was headers?

Reply:
> Shouldn't this also have a outlet.isAvailable condition like the block above?

`outlet.isAvailable` only has to be checked when there are trailers to be sent, so it happens a few lines below.

> What triggers `outlet.complete()` if there was headers?

Oooh good point. thanks.